text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,472 @@
+The FPC Shim fetches the state data from the peer and loads it into the Chaincode Enclave; then the FPC Shim verifies that the received data is correct (e.g. is actual committed data) with the help of the Ledger Enclave. This protects the Chaincode from various attacks that might be possible in the case of a compromised Peer. (An example and explanation is included below.) We refer to our paper (see prior art section) for more details on this type of attack.
+
+	When the particular chaincode function invocation finishes, the FPC Shim produces a cryptographic signature over the input arguments, the read-write set, and the result.
+	This is conceptually similar to the endorsement signature produced by the endorsing peer but instead of being rooted in an organizational entity like the peer it is based on the hardware and code entity and rooted in the attestation. Note that for this reason the FPC Shim keeps track of the read/writeset during the invocation inside the Chaincode Enclave. Then the FPC Shim returns the result along with the signature back to the Peer. The Peer then uses the FPC signature as endorsement signature for the proposal response and sends it back to the Client.
+
+* Step 3: Endorsement
+
+	![Endorsement](../images/fpc/high-level/FPC-Endorsement.png)
+
+	The Client receives the Proposal Response (and collects enough Proposal Responses from other endorsing peers to satisfy the chaincode’s Endorsement Policy). Moreover, the FPC Client SDK extension verifies that the proposal response signature has been produced by a "valid" Chaincode Enclave. If this verification step fails, there is no need for the client to proceed and the transaction invocation is aborted. Otherwise, the client continues and builds a transaction and submits it for ordering.
+
+* Step 4: Ordering
+
+	![Ordering](../images/fpc/high-level/FPC-Ordering.png)
+
+	With FPC we follow the normal Ordering service. While the Orderer is unmodified in FPC, the ordered transactions broadcast to the Peers in the Channel for validation now include attested endorsements. Note that we encourage the use of a BFT-based ordering service as FPC requires the Ordering service to be trusted.
+
+* Step 5: Validation
+
+	![Validate](../images/fpc/high-level/FPC-Validate.png)
+
+	As the Peers in the Channel receive the block of transactions, they perform the standard Fabric validation process.
+	In addition, a custom validation plugin in the Peer is responsible to verify FPC transactions. In particular, the FPC Validator queries its local FPC Registry to retrieve the enclave signature verification key and check that the signature was produced in an actual and correct Chaincode Enclave. This query retrieves the stored Attestation report associated with the Public Key of the Chaincode Enclave that produced and signed the transaction. The Attestation report is checked to verify the details of the Chaincode Enclave, affirming the validity of the Enclave.
```

Comment:
> In addition, a custom validation plugin in the Peer is responsible to verify FPC transactions. In particular, the FPC Validator queries its local FPC Registry to retrieve the enclave signature verification key and check that the signature was produced in an actual and correct Chaincode Enclave

It seems to me, that the validator needs to validate the entire block, to ensure that there is no omission of FPC transactions, no?

Reply:
so above: there are two places where that validation logic applies. By FPC Validator we are referring only to the plugin which is part of the (untrusted) legder, not the validation part of the trusted ledger.  

Note that security-wise there are two reasons to have the validator plugin: it (a) prevents a denial of service by ensuring the "normal" ledger considers a transaction also aborted when the trusted ledger faults some FPC transaction and (b) prevents unvalidated public FPC state from being wrongly commited (and potentially acted upon later).  The latter issue would go away but not supporting public state (i.e., `put_public_state`) but requiring a corresponding separate chaincode query which would decrypt the answer and return it as query result. I.e., this would be our fallback strategy iff validator (& endorsement) plugins would disappear unreplaced as part of the go-plugins eradication effort ...